### DISCUSSION TITLE
How to start a `MetaChain` server for debug?

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#191] [Q&A] [answered]
I have complete a temporary implement to update the `Group` on `MetaChain`, but I find that the `config.json` in the example is unmatch with the source code:

```
invalid config.json, err invalid type: map, expected an ObjectId at line 2 column 17
```

I has read the source code, I think it should be the format as follow:

```rust
#[derive(Serialize, Deserialize)]
pub struct MinerConfig {
    pub coinbase: ObjectId,
    pub interval: u32,
    pub bfc_spv_node: String,
    pub chain_type: Option<String>,
    pub bft_port: Option<u16>,
    pub bft_node_list: Option<Vec<String>>,
    pub miner_key_path: Option<String>,
    pub stat: Option<StatConfig>,
}
```

but, I don't known what does these fields mean, and how can I initialize them.

**Comment by weiqiushi** [comment, marked as answer]
## Compile the required binaries
Use the following command to compile the binaries needed to run the meta chain

`cargo build -p cyfs-meta-genesis`
Compile the meta chain genesis block creation tool

`cargo build -p cyfs-meta-miner`
Compile meta chain nodes

`cargo build -p cyfs-meta-spv`
Optionally, compile the official spv node of the meta chain. cyfs-ts-sdk depends on this service for some interfaces of the MetaChian class. the CYFS stack does not depend on this service

## Generate Genesis blocks
1. First prepare the Genesis private key, which can be generated by desc-tool tool, or directly with any public/private key pair of people/device
2. Prepare the Genesis configuration. Prepare a genesis.json file with the following content:
```json
{
  "chain_type": "standalone",
  "coinbase": "5aSixgLzAmyR5QbQibWFkrkNbBLagfawmK3pbdaYqyt6",
  "interval": 10,
  "bfc_spv_node": "http://127.0.0.1:11998",
  "coins": [
    {
      "coin_id": 0,
      "pre_balance": [["5aSixgLzAmyR5QbQibWFkrkNbBLagfawmK3pbdaYqyt6", 3000000000000]]
    }
  ],
  "price": {}
}
```
- chain_type: "standalone" means single node chain
- coinbase: coinbase of the block, we recommend using the object id of the Genesis private key.
- interval: interval between blocks. Unit is second.
- bfc_spv_node: current meaningless. Just keep this value. Cannot be deleted
- coins: the coins allocated by Genesis.
- price: currently meaningless, just keep it.
3. execute `cyfs-meta-genesis --config=genesis.json`, i.e. generate genesis block file in **current** directory

## Run the chain node
1. Prepare the chain configuration. Prepare a config.json file with something like:
```json
{
  "coinbase": "5aSixgLzAmyR5QbQibWFkrkNbBLagfawmK3pbdaYqyt6",
  "interval": 10,
  "bfc_spv_node": "http://127.0.0.1:11998",
  "stat": {
    "memory_stat": true,
	"storage": {
		"sqlite": {
			"path": "stat.db"
		}
	}
  }
}
```
where coinbase, interval, bfc_spv_node fields have the same meaning as above
- stat: The statistics database configuration of the chain. If you don't need the statistics of the chain, you can delete this field

Put this file in the Genesis block file directory
2. Execute `cyfs-meta-miner -p={genesis block directory}` to run the local test chain.
The default value of -p is the current directory

## Interact with the protocol stack and tools
The local configuration host: "127.0.0.1 nightly.meta.cyfs.com" will allow the nightly version of the stack and tools to interact with the local chain by default

**Reply by streetycat** [reply]
Thank you, I have build the `MetaChain` in `standalone` success follow your introduction. and edit some errors in it.

**Reply by streetycat** [reply]
by the way, we should remove follow files when we want to rebuild a new node.

```
block
db
state_db
state_db_${number}
```

**Reply by weiqiushi** [reply]
In fact, you'd better delete all the files and folders except for the binary and configuration files in order to start a new node

**Reply by streetycat** [reply]
> In fact, you'd better delete all the files and folders except for the binary and configuration files in order to start a new node

thank you, of course it's a good idea, but there are many other files my folder, and they flood these files.

you are a good man.
